pkg/server/plugin/nodeattestor/aws/client.go

```diff
@@ -78,17 +82,35 @@ func (cc *clientsCache) getClient(region string) (Client, error) {
 		return nil, status.Error(codes.FailedPrecondition, "not configured")
 	}
 
-	client, err := cc.newClient(cc.config, region)
+	var asssumeRoleArn bytes.Buffer
+	if cc.config.AssumeRoleArnTemplate != "" {
+		tmpl, err := template.New("assume-role-arn").Parse(cc.config.AssumeRoleArnTemplate)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to create template: %v", err)
+		}
+
+		data := struct {
+			AccountID string
+		}{
+			AccountID: accountID,
+		}
+
+		if err = tmpl.Execute(&asssumeRoleArn, data); err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to execute template: %v", err)
+		}
+	}
+
+	client, err := cc.newClient(cc.config, region, asssumeRoleArn.String())
 	if err != nil {
 		return nil, status.Errorf(codes.Internal, "failed to create client: %v", err)
 	}
 
-	cc.clients[region] = client
+	cc.clients[cacheKey] = client
 	return client, nil
 }
 
-func newClient(config *SessionConfig, region string) (Client, error) {
-	sess, err := newAWSSession(config.AccessKeyID, config.SecretAccessKey, region)
+func newClient(config *SessionConfig, region string, asssumeRoleARN string) (Client, error) {
+	sess, err := newAWSSession(config.AccessKeyID, config.SecretAccessKey, region, asssumeRoleARN)
```

Comment:
What happens if we assume a role arn that we already have? E.g. attesting an agent in the same AWS account as the SPIRE Server?

Reply:
I am pretty sure you will need to give additional permission to the current role to assume the same role. I will test this scenario and add the comment

Reply:
Ok great, thank you! If we can easily know the account that the code is running in, maybe it makes sense to detect this condition and not assume any role? Whatever you think makes the most sense is ok with me, just want to make sure that the case is handled well

Reply:
If both the spire server and the spire agent is running the same account and **assume_role** parameter is set, you see following error which make sense to me. (spire-server is the role name)

> ERRO[0034] Nodeattestor(aws_iid): failed to describe instance: AccessDenied: User: arn:aws:sts::<account-id>:assumed-role/spire-server/<aws-instance-id> is not authorized to perform: sts:AssumeRole on resource: arn:aws:iam::<account-id>:role/spire-server

I am leaning towards not making the special condition for the detecting the same account. Here are some considerations:
- Right now the spire server does not know its own account. We may need to make additional AWS calls to know the current account id
- If the use case does not need multiple accounts, you do not need to configure assume_role.
- If the use case needs multiple accounts, you will need to configure other accounts with additional permissions. In that case why not do the same on all accounts and keep the AWS configuration consistent across all accounts.

Reply:
Fair enough. I think the docs capture it. We can always add something later if we need to.